garrysmod/lua/includes/modules/redo.lua

```diff
@@ -0,0 +1,97 @@
+module( "redo", package.seeall )
+
+if CLIENT then
+
+	local redo_maxhistory = CreateClientConVar("cl_maxredohistory", "50", true, true)
+	
+	cvars.AddChangeCallback("cl_maxredohistory", function(name, old, new)
+		net.Start("redoLimit")
+		net.WriteInt(tonumber(new), 16)
+		net.SendToServer()
+	end)
+	
+	local function redo_message( len )
+		notification.AddLegacy( "Redone Item", 2, 3 )
+		surface.PlaySound( "buttons/button15.wav" )
+	end
+	
+	net.Receive( "redone", redo_message )
+	
+	local function redo_clear()
+		notification.AddLegacy( "Redo History Cleared", 4, 3 )
+		surface.PlaySound( "buttons/button14.wav" )
+	end
+	
+	net.Receive( "redoClear", redo_clear )
+	
+end
+
+if SERVER then
+
+	redo = {}
+	
+	local redohistory = CreateConVar( "sv_maxredohistory", "50", FCVAR_SERVER_CAN_EXECUTE, "Sets the maximum number of redos a client can elect to store." ) )
+
+	local function updatelimit( len, ply )
+		local max = math.Clamp(net.ReadInt(16), 0, math.min(redohistory:GetInt(), ply:GetInfoNum("cl_maxredohistory", 50))
+	
+		for i = 1, #redo[ply] - max do
```

Comment:
O(n^2) loop

http://wiki.garrysmod.com/page/table/remove
`table.remove` is an O(n) operation because it shifts the keys of all successive values one to the left. Walking backwards would solve the issue.